Lakeshore340Sup/lakeshore340excitations.c

```diff
@@ -0,0 +1,165 @@
+thresholdTempExcitationPair getLargestTempExcitationPairFromFileThatIsLessThanTempSp(FILE *thresholdsFile, epicsFloat64 tempSp) {
+    // Initialise values
+    thresholdTempExcitationPair lastTempExcitationPair = {DBL_MIN, -1};
+    char line[256];
+    // Loop through lines to find the first temp threshold that the setpoint is higher than, set the newExcitation accordingly
+    while (fgets(line, sizeof(line), thresholdsFile) != NULL) {
+        lastTempExcitationPair = getThresholdTempExcitationPairFromLine(line);
+        // Return the first valid pair that is less than or equal to the temperature setpoint
+        if (tempExcitationPairValid(lastTempExcitationPair) && tempSp < lastTempExcitationPair.temp) {
+            return lastTempExcitationPair;
+        }
+    }
+    // Return the last found pair or the original invalid pair
+    return lastTempExcitationPair;
+}
+
+// Set the excitation and temperature threshold pvs from the given thresholdPair
+long setThresholdPVs(aSubRecord *prec, thresholdTempExcitationPair thresholdPair) {
+    if (tempExcitationPairValid(thresholdPair)) {
+        // Set excitation and temp PVs
+        *(epicsEnum16*)prec->vala = thresholdPair.excitation;
+        *(epicsFloat64*)prec->valb = thresholdPair.temp;
+        // Set that the file has been found
+        *(epicsEnum16*)prec->valc = (epicsEnum16)0;
+        // Set that we should delay the change to wait for the temperature
+        *(epicsEnum16*)prec->vald = (epicsEnum16)1;
+        // Process output records
+        return 0;
+    } else {
+        errlogSevPrintf(errlogMajor, "No matching excitation found");
+        // Do not process output records
+        return -1;
+    }
+}
+
+bool containsInvalidLines(FILE *thresholdsFile) {
+    // Initialise values
+    thresholdTempExcitationPair lastTempExcitationPair = {DBL_MIN, -1}; 
+    bool invalidLines = false;
+    char line[256];
+    // Loop through lines to find the first temp threshold that the setpoint is higher than, set the newExcitation accordingly
+    while (fgets(line, sizeof(line), thresholdsFile) != NULL) {
+        lastTempExcitationPair = getThresholdTempExcitationPairFromLine(line);
+        // Return the first valid pair that is less than or equal to the temperature setpoint
+        if (!tempExcitationPairValid(lastTempExcitationPair)) {
+            errlogSevPrintf(errlogMajor, "Line Invalid:");
+            errlogSevPrintf(errlogMajor, line);
+            invalidLines = true;
+        }
+    }
+    // Reset file read to start
+    rewind(thresholdsFile);
+    // Return the last found pair or the original invalid pair
+    return invalidLines;
+}
+
+static long calculateNewExcitationFromThresholds(aSubRecord *prec) {
+    // Open file and check it has opened
+    FILE *thresholdsFile = fopen(prec->a, "r");
+    if (strstr(prec->a, "None.txt") != NULL) {
```

Comment:
I don't really like treating this one filename as special like this - I think it will lead to confusion as people will go looking for `None.txt` on the filesystem.

Maybe it would be better to have 2 separate options: the excitation file name, and whether to use it. This would be similar to how the ramping is done elsewhere.

Reply:
Replaced with a macro as suggested